app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClient.java

```diff
@@ -0,0 +1,451 @@
+     * so as not to contribute to battery drain.
+     */
+    static class OutputBuffer {
+
+        private static List<Event> QUEUE = new ArrayList<>();
+
+        /*
+         * When an item is added to QUEUE, wait this many ms before sending.
+         *
+         * If another item is added to QUEUE during this time, reset the
+         * countdown.
+         */
+        private static final int WAIT_MS = 30000;
+
+        /*
+         * When QUEUE.size() exceeds this value TIMER becomes non-interruptable.
+         */
+        private static final int MAX_QUEUE_SIZE = 128;
+
+        /*
+         * IMPLEMENTATION NOTE: Java Timer will provide the desired asynchronous
+         * countdown after a new item is added to QUEUE.
+         */
+        private static Timer TIMER = new Timer();
+
+        /*
+         * IMPLEMENTATION NOTE: Java abstract TimerTask class requires a run()
+         * method be defined.
+         *
+         * The run() method is called when the Timer expires.
+         */
+        private static class Task extends TimerTask {
+            public void run() {
+                sendAllScheduled();
+            }
+        }
+
+        /**
+         * If sending is enabled, dequeue and call send() on all scheduled items.
+         */
+        static synchronized void sendAllScheduled() {
+            TIMER.cancel();
+
+            if (ENABLED) {
+                /*
+                 * All items on QUEUE are permanently removed.
+                 */
+                send(QUEUE);
+                QUEUE.clear();
+            }
+        }
+
+        /**
+         * Schedule a request to be sent.
+         *
+         * @param event event data
+         */
+        static synchronized void schedule(Event event) {
+            /*
+             * Item is enqueued whether or not sending is enabled.
+             */
+            QUEUE.add(event);
+
+            if (ENABLED) {
+                if (QUEUE.size() >= MAX_QUEUE_SIZE) {
+                    /*
+                     * >= because while sending is disabled, any number of items
+                     * could be added to QUEUE without it emptying.
+                     */
+                    sendAllScheduled();
+                } else {
+                    /*
+                     * The arrival of a new item interrupts the timer and resets
+                     * the countdown.
+                     */
+                    TIMER.cancel();
+                    TIMER.purge();
+                    TIMER = new Timer();
+                    TIMER.schedule(new Task(), WAIT_MS);
+                }
+            }
+        }
+
+        /**
+         * If sending is enabled, attempt to send the provided events.
+         * Also batch the events ordered by their streams, as the QUEUE
+         * can contain events of different streams
+         * @param events list of events
+         */
+        static void send(List<Event> events) {
```

Comment:
It looks like we're not checking anywhere if the user has disabled Usage Reports in Preferences.  We should never be sending any kind of data if "send usage reports" is disabled.